core/trie/proof.go

```diff
@@ -0,0 +1,180 @@
+package trie
+
+import (
+	"fmt"
+
+	"github.com/NethermindEth/juno/core/crypto"
+	"github.com/NethermindEth/juno/core/felt"
+)
+
+// https://github.com/starknet-io/starknet-p2p-specs/blob/main/p2p/proto/snapshot.proto#L6
+type ProofNode struct {
+	Binary *Binary
+	Edge   *Edge
+}
+
+// Note: does not work for leaves
+func (pn *ProofNode) Hash() *felt.Felt {
+	switch {
+	case pn.Binary != nil:
+		return crypto.Pedersen(pn.Binary.LeftHash, pn.Binary.RightHash)
+	case pn.Edge != nil:
+		length := make([]byte, len(pn.Edge.Path.bitset))
+		length[len(pn.Edge.Path.bitset)-1] = pn.Edge.Path.len
+		pathFelt := pn.Edge.Path.Felt()
+		lengthFelt := new(felt.Felt).SetBytes(length)
+		return new(felt.Felt).Add(crypto.Pedersen(pn.Edge.Child, &pathFelt), lengthFelt)
+	default:
+		return nil
+	}
+}
+
+func (pn *ProofNode) PrettyPrint() {
+	if pn.Binary != nil {
+		fmt.Printf("  Binary:\n")
+		fmt.Printf("    LeftHash: %v\n", pn.Binary.LeftHash)
+		fmt.Printf("    RightHash: %v\n", pn.Binary.RightHash)
+	}
+	if pn.Edge != nil {
+		fmt.Printf("  Edge:\n")
+		fmt.Printf("    Child: %v\n", pn.Edge.Child)
+		fmt.Printf("    Path: %v\n", pn.Edge.Path)
+		fmt.Printf("    Value: %v\n", pn.Edge.Value)
+	}
+}
+
+type Binary struct {
+	LeftHash  *felt.Felt
+	RightHash *felt.Felt
+}
+
+type Edge struct {
+	Child *felt.Felt
+	Path  *Key
+	Value *felt.Felt
+}
+
+func isEdge(parentKey *Key, sNode storageNode) bool {
+	sNodeLen := sNode.key.len
+	if parentKey == nil { // Root
+		return sNodeLen != 0
+	}
+	return sNodeLen-parentKey.len > 1
+}
+
+// Note: we need to account for the fact that Junos Trie has nodes that are Binary AND Edge,
+// whereas the protocol requires nodes that are Binary XOR Edge
+func transformNode(tri *Trie, parentKey *Key, sNode storageNode) (*Edge, *Binary, error) {
+	isEdgeBool := isEdge(parentKey, sNode)
+
+	var edge *Edge
+	if isEdgeBool {
+		edgePath := path(sNode.key, parentKey)
+		edge = &Edge{
+			Path:  &edgePath,
+			Child: sNode.node.Value,
+		}
+	}
+	if sNode.key.len == tri.height { // Leaf
+		return edge, nil, nil
+	}
+	lNode, err := tri.GetNodeFromKey(sNode.node.Left)
+	if err != nil {
+		return nil, nil, err
+	}
+	rNode, err := tri.GetNodeFromKey(sNode.node.Right)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	rightHash := rNode.Value
+	if isEdge(sNode.key, storageNode{node: rNode, key: sNode.node.Right}) {
+		edgePath := path(sNode.node.Right, sNode.key)
+		rEdge := ProofNode{Edge: &Edge{
+			Path:  &edgePath,
+			Child: rNode.Value,
```

Comment:
To clarify, the rNode.Value is the hash?

Reply:
Yup